Port all bindings to use RunTimeTypedWrapper or RefCountedWrapper

andrewkaufman opened the pull request:
Not sure if it would have been better to make a ParameterWrapper and OpWrapper or something along those lines. Didn't seem like it would help too much though.

I also wasn't sure how to deal with the pure virtual methods (e.g. `OpWrap::doOperation`). The docs seem to say `isSubclassed()` should always be used, but then I found a few examples that weren't using it, mainly when dealing with pure virtual methods, so I followed that cue.

Note the last commit (da949e2) is the only one that required test changes, though I think they make sense if I understood the comments in `WrapperGarbageCollector`

johnhaddon commented on include/IECorePython/RefCountedBinding.h:
```diff
@@ -63,6 +63,27 @@ class RefCountedWrapper : public T, public WrapperGarbageCollector
 		template<typename Arg1, typename Arg2, typename Arg3>
 		RefCountedWrapper( PyObject *self, Arg1 arg1, Arg2 arg2, Arg3 arg3 );
 
+		template<typename Arg1, typename Arg2, typename Arg3, typename Arg4>
```
I believe that with C++11 we can do this properly with a single variadic template. Seems a pity to do it the hard way when we're about to make the switch...

andrewkaufman replied:
What would you think of merging as is and then I'll come back and clean it up when I add c++11 (in a couple days ideally)? I'm scrambling to get a bunch of things done this next week before I'm on holiday(ish) for a month, and it seems easier to bash ahead with something that's working.

andrewkaufman replied:
To be clear, I was the c++11 change until after I purge all the color/image code, to avoid having to patch up dead code.

andrewkaufman replied:
*delaying*

johnhaddon replied:
Sure, delaying is fine. There are a number of spots where we can apply this kind of simplfication, so they could all be done in one PR. Maybe add a todo to remind us?

johnhaddon commented on include/IECorePython/Wrapper.h:
```diff
@@ -1,96 +0,0 @@
-//////////////////////////////////////////////////////////////////////////
```
Minor thing, but the commit message should say "Use RefCounted*Wrapper* and RunTimeTyped*Wrapper* instead"

andrewkaufman replied:
As that'll require a rebase to fix, shall I also rebase in the ParameterWrapper stuff discussed above, or would you prefer that as separate new commits?

johnhaddon replied:
Rebasing it all in sounds good to me...

andrewkaufman commented:
I've addressed the comments hopefully. One thing I couldn't quite figure out is how to get `ModifyOp` (and derived classes) using `OpClass` / `OpWrapper`. This seemed to go smoothly for `Reader` -> `ReaderClass` / `ReaderWrapper`, but I couldn't get it working for `ModifyOp`... anyways, all tests are passing with this branch.

Here's a [temp commit](https://github.com/andrewkaufman/cortex/commit/d1ff39f499b253d47960dac1b4e12eaf5abdc916) (not included in this branch) of my feeble attempts to get `ModifyClass` working...

johnhaddon commented:
> Here's a temp commit (not included in this branch) of my feeble attempts to get ModifyClass working...

Were you thinking you'd like to get that working before we merge, or do you think we should consider it good enough for now? It's already an improvement on where we were, and ops aren't really the focus any more anyway...

andrewkaufman commented:
> Were you thinking you'd like to get that working before we merge, or do you think we should consider it good enough for now? It's already an improvement on where we were, and ops aren't really the focus any more anyway...

I was definitely hoping you'd say something like that and we'd just merge.
